runners/spark/src/test/java/org/apache/beam/runners/spark/SparkPortableExecutionTest.java

```diff
@@ -163,4 +168,75 @@ public void process(ProcessContext context) {
+   * of a whole executable stage is expensive and can cause unexpected behavior when the executable
+   * stage has side effects (BEAM-7131).
+   */
+  @Test(timeout = 120_000)
+  public void testParDoWithSideEffects() throws Exception {
+    PipelineOptions options = PipelineOptionsFactory.create();
+    options.setRunner(CrashingRunner.class);
+    options
+        .as(PortablePipelineOptions.class)
+        .setDefaultEnvironmentType(Environments.ENVIRONMENT_EMBEDDED);
+    Pipeline pipeline = Pipeline.create(options);
+    Path path = FileSystems.getDefault().getPath(TEST_DIR_PREFIX, UUID.randomUUID().toString());
+    File dir = new File(path.toString());
+    PCollection<String> a =
+        pipeline
+            .apply("impulse", Impulse.create())
+            .apply(
+                "A",
+                ParDo.of(
+                    new DoFn<byte[], String>() {
+                      @ProcessElement
+                      public void process(ProcessContext context) throws Exception {
+                        context.output("A");
+                        // ParDos A, B, and C will all be fused together into the same executable
+                        // stage. This check verifies that stage is not run more than once by
+                        // enacting a side effect via the local file system.
+                        Assert.assertFalse("ParDo A should only have been run once.", dir.exists());
+                        dir.mkdirs();
+                      }
+                    }));
+    PCollection<KV<String, String>> b =
+        a.apply(
+            "B",
+            ParDo.of(
+                new DoFn<String, KV<String, String>>() {
+                  @ProcessElement
+                  public void process(ProcessContext context) throws Exception {
+                    context.output(KV.of(context.element(), "B"));
+                  }
+                }));
+    PCollection<KV<String, String>> c =
+        a.apply(
+            "C",
+            ParDo.of(
+                new DoFn<String, KV<String, String>>() {
+                  @ProcessElement
+                  public void process(ProcessContext context) throws Exception {
+                    context.output(KV.of(context.element(), "C"));
+                  }
+                }));
+    // Use GBKs to force re-computation of executable stage unless cached.
+    b.apply(GroupByKey.create());
+    c.apply(GroupByKey.create());
+    RunnerApi.Pipeline pipelineProto = PipelineTranslation.toProto(pipeline);
+    JobInvocation jobInvocation =
+        SparkJobInvoker.createJobInvocation(
+            "testParDoWithSideEffects",
+            "testParDoWithSideEffectsRetrievalToken",
+            sparkJobExecutor,
+            pipelineProto,
+            options.as(SparkPipelineOptions.class));
+    jobInvocation.start();
+    while (jobInvocation.getState() != Enum.DONE) {
+      Thread.sleep(1000);
```

Comment:
Shall we use SameThreadExecutor[1] to avoid sleeping.

[1] https://google.github.io/guava/releases/17.0/api/docs/com/google/common/util/concurrent/MoreExecutors.html#sameThreadExecutor()